packages/service-library/tests/fastapi/test_lifespan_utils.py

```diff
@@ -38,3 +50,212 @@ async def cache_lifespan(app: FastAPI) -> AsyncIterator[State]:
     assert "setup CACHE" not in messages
     assert "shutdown  DB" in messages
     assert "shutdown CACHE" in messages
+
+
+@pytest.fixture
+def postgres_lifespan_mng() -> LifespanManager:
+    lifespan_manager = LifespanManager()
+
+    @lifespan_manager.add
+    async def _setup_postgres_sync_engine(_) -> AsyncIterator[State]:
+        with log_context(logging.INFO, "postgres_sync_engine"):
+            # pass state to children
+            yield {"postgres": {"engine": "Some Engine"}}
+
+    @lifespan_manager.add
+    async def _setup_postgres_async_engine(_, state: State) -> AsyncIterator[State]:
+        with log_context(logging.INFO, "postgres_async_engine"):
+            # pass state to children
+
+            state["postgres"].update(aengine="Some Async Engine")
+            yield state
+
+    return lifespan_manager
+
+
+@pytest.fixture
+def rabbitmq_lifespan() -> LifespanManager:
+    lifespan_manager = LifespanManager()
+
+    @lifespan_manager.add
+    async def _setup_rabbitmq(app: FastAPI) -> AsyncIterator[State]:
+        with log_context(logging.INFO, "rabbitmq"):
+
+            with pytest.raises(AttributeError, match="rabbitmq_rpc_server"):
+                _ = app.state.rabbitmq_rpc_server
+
+            # pass state to children
+            yield {"rabbitmq_rpc_server": "Some RabbitMQ RPC Server"}
+
+    return lifespan_manager
+
+
+async def test_app_lifespan_composition(
+    postgres_lifespan_mng: LifespanManager, rabbitmq_lifespan: LifespanManager
+):
+    # The app has its own database and rpc-server to initialize
+    # this is how you connect the lifespans pre-defined in servicelib
+
+    @postgres_lifespan_mng.add
+    async def setup_database(app: FastAPI, state: State) -> AsyncIterator[State]:
+
+        with log_context(logging.INFO, "app database"):
+            assert state["postgres"] == {
+                "engine": "Some Engine",
+                "aengine": "Some Async Engine",
+            }
+
+            with pytest.raises(AttributeError, match="database_engine"):
+                _ = app.state.database_engine
+
+            app.state.database_engine = state["postgres"]["engine"]
+
+            yield {}  # no update
+
+            # tear-down stage
+            assert app.state.database_engine
+
+    @rabbitmq_lifespan.add
+    async def setup_rpc_server(app: FastAPI, state: State) -> AsyncIterator[State]:
+        with log_context(logging.INFO, "app rpc-server"):
+            assert "rabbitmq_rpc_server" in state
+
+            app.state.rpc_server = state["rabbitmq_rpc_server"]
+
+            yield {}
+
+    # Composes lifepans
+    app_lifespan = LifespanManager()
+    app_lifespan.include(postgres_lifespan_mng)
+    app_lifespan.include(rabbitmq_lifespan)
```

Comment:
Interesting, had no idea this worked like this. I spent quite some time with them but could not figure it out.

Q: what if you set something in the state of lifespana1, can you access it in the lifespan2, did you test this? I had trouble with it.

Reply:
yes, we pass states from one to the other. Check tte associated checks